lms/static/scripts/frontend_apps/components/SubmitGradeForm.tsx

```diff
@@ -84,6 +84,23 @@ export default function SubmitGradeForm({
   // Used to handle keyboard input changes for the grade input field.
   const inputRef = useRef<HTMLInputElement | null>(null);
 
+  // Listen for page unloads to make sure instructors do not lose unsaved data
+  useEffect(() => {
+    const listener = (e: BeforeUnloadEvent) => {
+      // If grade was saved or not changed at all, allow to safely exit
+      if (gradeSaved || grade.data === inputRef.current?.value) {
```

Comment:
The `gradeSaved` state is potentially a bit confusing because it doesn't represent "is the displayed grade value saved", since it starts out as _false_. Instead it just exists to trigger an animation. Is `grade.data` updated after a grade is submitted? If so then is that test sufficient on its own?

Reply:
I have refactored to use a new ref which tracks if the grade is unsaved.

* It is initially false (so the grade is saved).
* It is set to false also after submitting the grade.
* Every time the grade input value changes, we compare it with the originally loaded grade in order to know if it's unsaved.

This seems to work, as it does not cause re-renders, and tests pass or fail consistently. Using `gradeSaved` it was a bit inconsistent and sometimes the test could fail.

Reply:
> What stops us from fixing that?

`grade` is the result of `useFetch`, and currently its fetcher callback is invoked only if the student changes.

I thought this was by design (as in, let's save one API call), but we could add some other dependency that forces it to re-fetch after submitting the grade.

Do you think it would be acceptable to do that extra fetch call? Or do you think we should instead derive some other piece of state based on loaded grade + saved grade?

Reply:
> grade is the result of useFetch, and currently its fetcher callback is invoked only if the student changes.

Ah, OK. It would make sense for `useFetch` to have a way to replace the result without needing to make a call to the backend. I added it in https://github.com/hypothesis/lms/pull/5677.